js/package.json

```
@@ -76,13 +78,23 @@
     "topojson": "^1.6.24",
     "underscore": "^1.8.3"
   },
+  "jupyterlab": {
+    "extension": "lib/jupyterlab-plugin",
+    "outputDir": "../share/jupyter/labextensions/bqplot",
+    "sharedPackages": {
+      "@jupyter-widgets/base": {
+        "bundled": false,
+        "singleton": true
+      },
+      "threejs": {
+        "bundled": true
+      }
```

```suggestion
      }
```
The default is to bundle dependencies, so this is not needed, and the npm package name is `three`, so this isn't doing anything anyway.

---

Maybe this should be `"three": { "singleton": true }` ?

---

If multiple extensions each provide a copy of three, setting the singleton to true just makes sure they all use the same version (which would be the highest version), even if is the wrong version.

I think it's fine to just not put anything, and let the system take care of version conflicts (i.e., things will be shared when possible, but if not, then multiple versions can coexist) - unless it is really important to share the same module even if it is the wrong version. Or if three itself cannot handle having multiple versions on a page.

---

If multiple extensions need to pass references to three objects to each other, then it is important that they are shared, but I guess that should be managed via semver then, since other extensions might want to use a different version of threejs independently of what the other two are doing.

---

FYI, here is my analysis of how these options interplay: https://github.com/jupyterlab/jupyterlab/issues/9496#issuecomment-749381599